File: aiquota/pace.py
from aiquota.models import PaceResult, QuotaWindow

STABLE_FRACTION = 0.05
EXHAUSTED_PERCENT = 100.0


def is_exhausted(w: QuotaWindow) -> bool:
    return w.used_percent >= EXHAUSTED_PERCENT


def compute_pace(w: QuotaWindow) -> PaceResult | None:
    elapsed_secs = w.window_seconds - w.reset_seconds
    elapsed_frac = elapsed_secs / w.window_seconds
    expected_percent = elapsed_frac * 100
    deviation = w.used_percent - expected_percent
    projected_at_reset = None
    seconds_to_exhaust = None
    if elapsed_secs > 0 and w.used_percent > 0:
        rate_per_sec = w.used_percent / elapsed_secs
        seconds_to_exhaust = (100 - w.used_percent) / rate_per_sec
        projected_at_reset = w.used_percent + rate_per_sec * w.reset_seconds
    stable = STABLE_FRACTION < elapsed_frac < 1 - STABLE_FRACTION
    return PaceResult(
        deviation=deviation, projected_at_reset=projected_at_reset, seconds_to_exhaust=seconds_to_exhaust, stable=stable
    )


# Pace deviation thresholds (percentage points: used% - expected%).
PACE_COOL_BELOW = -10
PACE_WARN_ABOVE = 5
PACE_HOT_ABOVE = 15
SHORT_WIN_HOT_PERCENT = 85

TINT_RANK = {"unknown": 0, "stale": 0, "ok": 1, "cool": 1, "warn": 2, "hot": 3}


def tint_for(pace: PaceResult | None, used_percent: float | None, *, is_short: bool) -> str:
    if used_percent is None:
        return "unknown"
    if is_short and used_percent >= SHORT_WIN_HOT_PERCENT:
        return "hot"
    if not pace or not pace.stable:
        if used_percent >= 95:
            return "hot"
        if used_percent >= 80:
            return "warn"
        return "ok"
    if pace.deviation >= PACE_HOT_ABOVE:
        return "hot"
    if pace.deviation >= PACE_WARN_ABOVE:
        return "warn"
    if pace.deviation <= PACE_COOL_BELOW:
        return "cool"
    return "ok"


def binding_tint(short_tint: str, long_tint: str) -> str:
    if short_tint == "hot":
        return "hot"
    return short_tint if TINT_RANK.get(short_tint, 0) > TINT_RANK.get(long_tint, 0) else long_tint

File: aiquota/frontend/pace.ts
/**
 * Pace math for the browser dashboard — the TypeScript half of `aiquota/pace.py`.
 *
 * The API serves a snapshot; the reset countdown, and therefore the elapsed fraction every
 * number here derives from, keeps moving between snapshots. So this runs client-side, exactly
 * as it does in the GNOME extension, rather than being served pre-computed and going stale
 * between polls.
 *
 * `aiquota/pace.py` is the canonical statement of the thresholds and the derivation (the
 * reasoning is in `aiquota/gnome/DESIGN.md` § Pace math). `format.test.ts` holds this port to
 * the strings that module produces for the shared scenarios, so the two cannot disagree about
 * a pace, a forecast, or a tint without failing.
 */

export type Tint = "cool" | "ok" | "warn" | "hot" | "unknown" | "stale";

export type Pace = {
  deviation: number;
  projectedAtReset: number | null;
  secondsToExhaust: number | null;
  /** Pace is meaningless in the first and last twentieth of a window; too little has happened. */
  stable: boolean;
};

export type WindowMath = { usedPercent: number; resetSeconds: number; windowSeconds: number };

const STABLE_FRACTION = 0.05;
const EXHAUSTED_PERCENT = 100;
const PACE_COOL_BELOW = -10;
const PACE_WARN_ABOVE = 5;
const PACE_HOT_ABOVE = 15;
const SHORT_WIN_HOT_PERCENT = 85;

const TINT_RANK: Record<Tint, number> = { unknown: 0, stale: 0, ok: 1, cool: 1, warn: 2, hot: 3 };

export function isExhausted(window: WindowMath): boolean {
  return window.usedPercent >= EXHAUSTED_PERCENT;
}

export function elapsedFraction(window: WindowMath): number {
  return clamp01((window.windowSeconds - window.resetSeconds) / window.windowSeconds);
}

export function computePace(window: WindowMath): Pace {
  const elapsedSeconds = window.windowSeconds - window.resetSeconds;
  const elapsed = elapsedSeconds / window.windowSeconds;
  const spending = elapsedSeconds > 0 && window.usedPercent > 0;
  const ratePerSecond = window.usedPercent / elapsedSeconds;
  return {
    deviation: window.usedPercent - elapsed * 100,
    projectedAtReset: spending ? window.usedPercent + ratePerSecond * window.resetSeconds : null,
    secondsToExhaust: spending ? (100 - window.usedPercent) / ratePerSecond : null,
    stable: elapsed > STABLE_FRACTION && elapsed < 1 - STABLE_FRACTION,
  };
}

/**
 * `isShort`: this is not the provider's longest window. A short window at 85% is the
 * immediate-pain case — the next burst gets 429s — whatever its pace says.
 */
export function tintFor(pace: Pace | null, usedPercent: number, { isShort }: { isShort: boolean }): Tint {
  if (isShort && usedPercent >= SHORT_WIN_HOT_PERCENT) return "hot";
  if (pace === null || !pace.stable) {
    if (usedPercent >= 95) return "hot";
    if (usedPercent >= 80) return "warn";
    return "ok";
  }
  if (pace.deviation >= PACE_HOT_ABOVE) return "hot";
  if (pace.deviation >= PACE_WARN_ABOVE) return "warn";
  if (pace.deviation <= PACE_COOL_BELOW) return "cool";
  return "ok";
}

/** The tint a whole provider takes: the most urgent of its windows'. */
export function bindingTint(tints: Tint[]): Tint {
  return tints.reduce((worst, tint) => (TINT_RANK[tint] > TINT_RANK[worst] ? tint : worst), "unknown");
}

function clamp01(value: number): number {
  return Math.max(0, Math.min(1, value));
}
